classes/SpecificPrice.php

```diff
@@ -337,24 +383,148 @@ protected static function formatIntInQuery($first_value, $second_value)
         }
     }
 
-    public static function getSpecificPrice($id_product, $id_shop, $id_currency, $id_country, $id_group, $quantity, $id_product_attribute = null, $id_customer = 0, $id_cart = 0, $real_quantity = 0)
+    /**
+     * Check if the given product could have a specific price
+     *
+     * @param $idProduct
+     *
+     * @return bool
+     */
+    protected static function couldHaveSpecificPrice($idProduct)
+    {
+        if (SpecificPrice::$_hasGlobalProductRules === null) {
+            $queryHasGlobalRule = 'SELECT 1 FROM `' . _DB_PREFIX_ . 'specific_price` WHERE id_product = 0';
+            $row = Db::getInstance(_PS_USE_SQL_SLAVE_)->getRow($queryHasGlobalRule);
+            SpecificPrice::$_hasGlobalProductRules = !empty($row);
+        }
+        if (SpecificPrice::$_hasGlobalProductRules) {
+            return true;
+        }
+
+        if (!array_key_exists($idProduct, SpecificPrice::$_couldHaveSpecificPriceCache)) {
+            $query = 'SELECT 1 FROM `' . _DB_PREFIX_ . 'specific_price` WHERE id_product = ' . (int)$idProduct;
+            $row = Db::getInstance(_PS_USE_SQL_SLAVE_)->getRow($query);
+            SpecificPrice::$_couldHaveSpecificPriceCache[$idProduct] = !empty($row);
+        }
+
+        return SpecificPrice::$_couldHaveSpecificPriceCache[$idProduct];
+    }
+
+    /**
+     * Compute the cache key by setting to 0 the fields which doesn't have any specific values in the DB
+     *
+     * @param int $id_product
+     * @param int $id_shop
+     * @param int $id_currency
+     * @param int $id_country
+     * @param int $id_group
+     * @param int $quantity
+     * @param int $id_product_attribute
+     * @param int $id_customer
+     * @param int $id_cart
+     * @param int $real_quantity
+     *
+     * @return string
+     */
+    protected static function computeKey(
+        $id_product,
+        $id_shop,
+        $id_currency,
+        $id_country,
+        $id_group,
+        $quantity,
+        $id_product_attribute,
+        $id_customer,
+        $id_cart,
+        $real_quantity
+    )
+    {
+        if (self::$_no_specific_values !== null) {
+            // $_no_specific_values contains the fieldName from the DB which don't have values != 0
+            // So it's ok the set the value to 0 for those fields to improve the cache efficiency
+            // Note that the variableName from the DB needs to match the function args name
+            // I.e. if the computeKey args are converted at some point in camelCase, we will need to introduce a
+            // snakeCase to camelCase conversion of $variableName
+            foreach (array_keys(self::$_no_specific_values) as $variableName) {
```

Comment:
That's actually part of the bigger issue of "is there a point to allowing overrides really?"